pkg/progress/progress.go

```diff
@@ -0,0 +1,164 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package progress
+
+import (
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/rs/zerolog/log"
+)
+
+var Channel = make(chan ProgressParams, 10000)
+var CancelProgressChannel = make(chan ProgressToken, 10000)
+
+type Tracker struct {
+	channel              chan ProgressParams
+	cancelChannel        chan ProgressToken
+	token                ProgressToken
+	cancellable          bool
+	lastReport           time.Time
+	lastReportPercentage int
+	finished             bool
+}
+
+func NewTestTracker(channel chan ProgressParams, cancelChannel chan ProgressToken) *Tracker {
+	return &Tracker{
+		channel:       channel,
+		cancelChannel: cancelChannel,
+		// deepcode ignore HardcodedPassword: false positive
+		token:                "token",
+		cancellable:          true,
+		lastReportPercentage: -1,
+	}
+}
+
+func NewTracker(cancellable bool) *Tracker {
+	return &Tracker{
+		channel:       Channel,
+		cancelChannel: CancelProgressChannel,
+		cancellable:   cancellable,
+		finished:      false,
+	}
+}
+
+func (t *Tracker) BeginUnquantifiableLength(title, message string) {
+	t.begin(title, message, true)
+}
+
+func (t *Tracker) begin(title string, message string, unquantifiableLength bool) {
+	params := newProgressParams(title, message, t.cancellable, unquantifiableLength)
+	t.token = params.Token
+
+	t.send(ProgressParams{
+		Token: t.token,
+		Value: nil,
+	})
+
+	t.send(params)
+	t.lastReport = time.Now()
+}
+
+func (t *Tracker) Begin(title string) {
+	t.begin(title, "", false)
+}
+
+func (t *Tracker) BeginWithMessage(title, message string) {
+	t.begin(title, message, false)
+}
+
+func (t *Tracker) ReportWithMessage(percentage int, message string) {
+	if time.Now().Before(t.lastReport.Add(time.Second)) || percentage <= t.lastReportPercentage {
+		return
+	}
+	progress := ProgressParams{
+		Token: t.token,
+		Value: WorkDoneProgressReport{
+			WorkDoneProgressKind: WorkDoneProgressKind{Kind: "report"},
+			Percentage:           percentage,
+			Message:              message,
+		},
+	}
+	t.send(progress)
+	t.lastReport = time.Now()
+	t.lastReportPercentage = percentage
+}
+
+func (t *Tracker) Report(percentage int) {
+	t.ReportWithMessage(percentage, "")
+}
+
+func (t *Tracker) End() {
+	t.EndWithMessage("")
+}
+
+func (t *Tracker) EndWithMessage(message string) {
+	if t.finished {
+		panic("Called end progress twice. This breaks LSP in Eclipse fix me now and avoid headaches later")
```

Comment:
*cough*

Reply:
oops

Reply:
Is this something we should fix now? It's a bit annoying to mention LSP and Eclipse, plus not ideal to cause a panic in the CLI although if the function is used correctly we shouldn't cause it. It might be better to log an error instead and return early maybe?